File: 2021/fsharp/Aoc2021/Day06.fs
module Day06

open Utils

let simulateFishPopulation (fishTimers: int list) =
    let timers =
        fishTimers
        |> List.countBy (fun timer -> timer)
        |> List.fold
            (fun (timers: int64 []) (timer, count) ->
                timers.[timer] <- count
                timers)
            [| 0L; 0L; 0L; 0L; 0L; 0L; 0L; 0L; 0L |]

    let fishCount timers = timers |> Array.sum

    seq {
        while true do
            yield timers |> fishCount

            let motherCount = timers.[0]
            Array.blit timers 1 timers 0 8
            timers.[8] <- motherCount
            timers.[6] <- timers.[6] + motherCount
    }

let input =
    "06"
    |> readInputText
    |> String.split ','
    |> Array.map int
    |> Array.toList

let job1 () =
    input
    |> simulateFishPopulation
    |> Seq.item 80
    |> string

let job2 () =
    input
    |> simulateFishPopulation
    |> Seq.item 256
    |> string

File: 2021/csharp/Aoc2021/Day06.cs
namespace Aoc2021;

public class Day06 : IDay
{
    public override string Day { get; } = nameof(Day06)[3..];

    public override string Result1()
    {
        return SimulateFishPopulation(InputText.Split(',').Select(ParseLong).ToList())
            .ElementAt(80)
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return SimulateFishPopulation(InputText.Split(',').Select(ParseLong).ToList())
            .ElementAt(256)
            .ToString(CultureInfo.InvariantCulture);
    }

    public static IEnumerable<long> SimulateFishPopulation(List<long> fishTimers)
    {
        yield return fishTimers.Count;

        long[] timers = fishTimers
            .Aggregate(
                new long[] { 0, 0, 0, 0, 0, 0, 0, 0, 0 },
                (timers, timer) =>
                {
                    timers[timer] += 1;
                    return timers;
                }
            );

        while (true)
        {
            long motherCount = timers[0];
            for (long i = 1; i != timers.Length; i++)
            {
                timers[i - 1] = timers[i];
            }
            timers[8] = motherCount;
            timers[6] += motherCount;

            yield return timers.Sum();
        }
    }
}
